flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java

```diff
@@ -0,0 +1,155 @@
+                                .orElse("(null)")));
+
+        final Set<Execution> concurrentlyFailedExecutions =
+                failureHandlingResult.getVerticesToRestart().stream()
+                        .filter(
+                                e ->
+                                        !e.equals(
+                                                failureHandlingResult
+                                                        .getExecutionVertexIdOfFailedTask()
+                                                        .orElse(null)))
+                        .map(executionVertexLookup)
+                        .map(ExecutionVertex::getCurrentExecutionAttempt)
+                        .filter(e -> e.getFailureInfo().isPresent())
+                        .collect(Collectors.toSet());
+
+        return new FailureHandlingResultSnapshot(
+                rootCauseExecution,
+                ErrorInfo.handleMissingThrowable(failureHandlingResult.getError()),
+                failureHandlingResult.getTimestamp(),
+                concurrentlyFailedExecutions);
+    }
+
+    @VisibleForTesting
+    FailureHandlingResultSnapshot(
+            @Nullable Execution rootCauseExecution,
+            Throwable rootCause,
+            long timestamp,
+            Set<Execution> concurrentlyFailedExecutions) {
+        Preconditions.checkArgument(
+                rootCauseExecution == null
+                        || !concurrentlyFailedExecutions.contains(rootCauseExecution),
+                "The rootCauseExecution should not be part of the concurrentlyFailedExecutions map.");
+
+        this.rootCauseExecution = rootCauseExecution;
+        this.rootCause = Preconditions.checkNotNull(rootCause);
+        this.timestamp = timestamp;
+        this.concurrentlyFailedExecutions =
+                Preconditions.checkNotNull(concurrentlyFailedExecutions);
+    }
+
+    /**
+     * Returns the {@link Execution} that handled the root cause for this failure. An empty {@code
+     * Optional} will be returned if it's a global failure.
+     *
+     * @return The {@link Execution} that handled the root cause for this failure.
+     */
+    public Optional<Execution> getRootCauseExecution() {
+        return Optional.ofNullable(rootCauseExecution);
+    }
+
+    /**
+     * The actual failure that is handled.
+     *
+     * @return The {@code Throwable}.
+     */
+    public Throwable getRootCause() {
+        return rootCause;
+    }
+
+    /**
+     * The time the failure occurred.
+     *
+     * @return The time of the failure.
+     */
+    public long getTimestamp() {
+        return timestamp;
+    }
+
+    /**
+     * All {@link Execution Executions} that failed and are planned to be restarted as part of this
+     * failure handling.
+     *
+     * @return The concurrently failed {@code Executions}.
+     */
+    public Iterable<Execution> getConcurrentlyFailedExecution() {
+        return concurrentlyFailedExecutions;
```

Comment:
I would suggest to return `Collections.unmodifiableSet(concurrentlyFailedExecutions)`.